domain/user/service/service.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"github.com/juju/juju/core/user"
+	usererrors "github.com/juju/juju/domain/user/errors"
+)
+
+// State describes retrieval and persistence methods for upgrade info.
+type State interface {
+	// AddUser will add a new user to the database. If the user already exists
+	// an error that satisfies usererrors.AlreadyExists will be returned.
+	AddUser(context.Context, string, string, string, string) error
+
+	// GetUser will retrieve the user specified by name from the database where
+	// the user is active and has not been deleted. If the user does not exist
+	// or is deleted an error that satisfies usererrors.NotFound will be
+	// returned.
+	GetUser(context.Context, string) (user.User, error)
+	// RemoveUser marks the user as deleted. This obviates the ability of a user
+	// to function, but keeps the user retaining provenance, i.e. auditing.
+	RemoveUser(context.Context, string) error
+	// SetPassword removes any active activation keys and sets the user password
+	SetPassword(context.Context, string, user.UserPassword) error
+	// GenerateUserActivationKey makes a new activation key for the given user.
+	// It replaces any existing key for the user.
+	GenerateUserActivationKey(context.Context, string) error
+	// RemoveUserActivationKey removes any activation key ass the given user
+	RemoveUserActivationKey(context.Context, string) error
```

Comment:
I don't think we need this func just yet as this operation should happen with set password using activation key as describe above. We ideally want to get all of these operations in the one transaction.

Reply:
Done & Done